ext/ldap/ldap.c

```diff
@@ -397,7 +398,6 @@ static int _php_ldap_control_from_array(LDAP *ld, LDAPControl** ctrl, zval* arra
 						if (ber_flatten2(vrber, control_value, 0) == -1) {
 							rc = -1;
 						}
-						ber_free(vrber, 1);
```

Comment:
Completely glossed over this line before ... this doesn't look right to me. Why does removing this not cause a memory leak? I don't see anything else that would free vrber.

Reply:
Somehow this caused a "Protocol error" which appeared to be a bad memory access (access to freed memory). I guess, this is specific to how the ``ber_flatten2()`` is [implemented](https://github.com/openldap/openldap/blob/947bbfbf5ac6ee21508d9e67dddeb124fcf8da54/libraries/liblber/io.c#L385). If the third argument called ``alloc`` is zero, then it just assigns ``vrber->ber_buf``` to ``control_value->ber_buf``. The memory is released later with ``ber_memfree(control_value)`` at [line 618](https://github.com/php/php-src/blob/265c676142ca9ad76853e3eca86d4631377a2ae5/ext/ldap/ldap.c#L618). From what I've seen, this is recurring pattern in ``ldap.c`` - I couldn't find other pieces of code with ``ber_free()`` after ``ber_flatten2()``.

Reply:
Thanks for the explanation, that makes sense!

Reply:
Hm... just got back to this, and it looks like I could be wrong. The memory allocated for ``vrber->ber_buf`` is *probably* freed with ``ber_memfree(control_value)``, but there is still memory allocated for ``vrber`` itself. So removing completely this line was not right, perhaps we should replace it with ``ber_free(vrber, 0)``, or move the ``ber_free(vrber, 1)`` near the end of the function.